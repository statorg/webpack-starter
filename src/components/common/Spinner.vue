<template>
  <div class="spinner" :style="styles">
    <svg
      xmlns="http://www.w3.org/2000/svg"
      :viewBox="viewBox"
      class="spinner__svg"
    >
      <circle
        fill="transparent"
        :cx="viewBoxSize"
        :cy="viewBoxSize"
        :r="radius"
        :stroke-width="strokeWidth"
        :stroke-dasharray="strokeDashArray"
        :stroke-dashoffset="strokeDashOffset"
        class="spinner__circle"
      />
    </svg>
  </div>
</template>

<script>
export default {
  name: 'Spinner',
  props: {
    size: {
      type: [Number, String],
      default: 32
    },
    width: {
      type: [Number, String],
      default: 4
    }
  },
  data () {
    return {
      radius: 20
    }
  },
  computed: {
    styles () {
      return {
        width: `${this.size}px`,
        height: `${this.size}px`
      }
    },
    circumference () {
      return 2 * Math.PI * this.radius
    },
    strokeDashArray () {
      return Math.round(this.circumference * 1000) / 1000
    },
    strokeDashOffset () {
      return `${this.circumference}px`
    },
    strokeWidth () {
      return Number(this.width) / +this.size * this.viewBoxSize * 2
    },
    viewBoxSize () {
      return this.radius / (1 - Number(this.width) / +this.size)
    },
    viewBox () {
      return `0 0 ${this.viewBoxSize * 2} ${this.viewBoxSize * 2}`
    }
  }
}
</script>

<style lang="scss">
.spinner {
  position: relative;
  display: inline-flex;
  vertical-align: middle;
  justify-content: center;
  align-items: center;

  &__svg {
    width: 100%;
    height: 100%;
    margin: auto;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    z-index: 0;
    animation: progress-circular-rotate 1.4s linear infinite;
    transform-origin: center center;
    transition: all 0.2s ease-in-out;

    @keyframes progress-circular-rotate {
      100% {
        transform: rotate(360deg);
      }
    }
  }

  &__circle {
    stroke: currentColor;
    z-index: 2;
    transition: all 0.6s ease-in-out;
    animation: progress-circular-dash 1.4s ease-in-out infinite;
    stroke-linecap: round;
    stroke-dasharray: 80, 200;
    stroke-dashoffset: 0;

    @keyframes progress-circular-dash {
      0% {
        stroke-dasharray: 1, 200;
        stroke-dashoffset: 0;
      }

      50% {
        stroke-dasharray: 100, 200;
        stroke-dashoffset: -15px;
      }

      100% {
        stroke-dasharray: 100, 200;
        stroke-dashoffset: -125px;
      }
    }
  }
}
</style>
