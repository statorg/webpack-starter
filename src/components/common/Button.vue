<script>
import Spinner from './Spinner'

export default {
  name: 'Button',
  functional: true,
  props: {
    block: Boolean,
    size: {
      type: String,
      default: ''
    },
    variant: {
      type: String,
      default: ''
    },
    loading: Boolean
  },
  render (h, { props, data, slots }) {
    const componentData = {
      class: {
        [`button_${props.variant}`]: !!props.variant,
        [`button_${props.size}`]: !!props.size,
        button_block: props.block,
        button_loading: props.loading
      }
    }

    let tag = 'button'
    if (data.attrs.href) {
      tag = 'a'
      delete data.attrs.type
    } else {
      data.attrs.type = data.attrs.type || 'button'
      delete data.attrs.href
    }

    return (
      <tag class="button" {...data} {...componentData}>
        <span class="button__content">{slots().default}</span>
        {
          props.loading &&
          <Spinner class="button__spinner" size={23} width={2} />
        }
      </tag>
    )
  }
}
</script>

<style lang="scss">
.button {
  cursor: pointer;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  vertical-align: middle;
  text-align: center;
  color: $body-color;
  user-select: none;
  background-color: transparent;
  border: 1px solid transparent;
  line-height: $line-height;
  font-size: 16px;
  padding: 6px 12px;
  border-radius: 4px;
  transition-duration: 0.2s; // 0.28s
  transition-property: background-color, border-color, box-shadow, transform, opacity;
  transition-timing-function: easing(fast-out-slow-in);
  max-width: 100%;
  flex: 0 0 auto;
  position: relative;

  &:hover {
    text-decoration: none;
  }

  &[disabled] {
    pointer-events: none;
    opacity: 0.65;
  }

  &_primary {
    color: #ffffff;
    background-color: $primary;
    border-color: $primary;

    &:hover {
      background-color: lighten($primary, 4%);
      border-color: lighten($primary, 4%);
    }

    &:active {
      background-color: darken($primary, 4%);
      border-color: darken($primary, 4%);
    }
  }

  &_small {
    padding: 4px 8px;
    font-size: 14px;
    border-radius: 3px;
  }

  &_large {
    padding: 8px 16px;
    font-size: 20px;
    border-radius: 5px;
  }

  &_block {
    display: flex;
    width: 100%;
  }

  &_loading {
    pointer-events: none;
    transition: none;
  }

  &__content {
    align-items: center;
    display: flex;
    flex: 1 0 auto;
    justify-content: inherit;
  }

  &_loading &__content {
    opacity: 0;
  }

  &__spinner.spinner {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }
}

@include vue(v-button) {
  @extend .button;
  -webkit-appearance: none !important;

  &[variant="primary"] {
    @extend .button_primary;
  }

  &[size="small"] {
    @extend .button_small;
  }

  &[size="large"] {
    @extend .button_large;
  }

  &[block] {
    @extend .button_block;
  }
}
</style>
