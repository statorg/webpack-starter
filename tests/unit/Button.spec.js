import { shallowMount } from '@vue/test-utils'
import Button from '@/components/common/Button'
import Spinner from '@/components/common/Spinner'

describe('Button.vue', () => {
  it('should have button_primary class when variant prop is passed as primary', () => {
    const wrapper = shallowMount(Button, {
      propsData: {
        variant: 'primary'
      }
    })
    expect(wrapper.classes()).toContain('button_primary')
  })

  it('should emit click event when clicked', () => {
    const onClick = jest.fn()
    const wrapper = shallowMount(Button, {
      listeners: {
        click: onClick
      }
    })
    wrapper.trigger('click')
    expect(onClick).toHaveBeenCalled()
    expect(onClick.mock.calls[0][0]).toBeInstanceOf(MouseEvent)
  })

  it('should render with default slot', () => {
    const text = 'Button text'
    const wrapper = shallowMount(Button, {
      slots: {
        default: text
      }
    })
    expect(wrapper.text()).toContain(text)
  })

  it('should render with spinner when loading prop is passed as true', () => {
    const wrapper = shallowMount(Button, {
      propsData: {
        loading: true
      }
    })
    expect(wrapper.contains(Spinner)).toBe(true)
  })

  it('should render an <a> tag when href attribute is passed', () => {
    const wrapper = shallowMount(Button, {
      attrs: {
        href: 'http://www.google.com'
      }
    })
    expect(wrapper.is('a')).toBe(true)
  })

  it('should render a <button> tag when href attribute is not passed', () => {
    const wrapper = shallowMount(Button)
    expect(wrapper.is('button')).toBe(true)
  })

  it('should render and match snapshot', () => {
    const wrapper = shallowMount(Button)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
