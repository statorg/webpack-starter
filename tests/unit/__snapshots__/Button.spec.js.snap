// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.vue should render and match snapshot 1`] = `<button type="button" class="button"><span class="button__content"></span></button>`;
